mmv1/third_party/terraform/utils/test_utils.go

```diff
@@ -156,3 +159,67 @@ func checkDataSourceStateMatchesResourceStateWithIgnores(dataSourceName, resourc
 		return nil
 	}
 }
+
+// testExtractResourceAttr navigates a test's state to find the specified resource (or data source) attribute and makes the value
+// accessible via the attributeValue string pointer.
+func testExtractResourceAttr(resourceName string, attributeName string, attributeValue *string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resourceName] // To find a datasource, include `data.` at the start of the resourceName value
+
+		if !ok {
+			return fmt.Errorf("resource name %s not found in state", resourceName)
+		}
+
+		attrValue, ok := rs.Primary.Attributes[attributeName]
+
+		if !ok {
+			return fmt.Errorf("attribute %s not found in resource %s state", attributeName, resourceName)
+		}
+
+		*attributeValue = attrValue
+
+		return nil
+	}
+}
+
+// testCheckAttributeValuesEqual compares two string pointers, which have been used to retrieve attribute values from the test's state.
+func testCheckAttributeValuesEqual(i *string, j *string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if testStringValue(i) != testStringValue(j) {
+			return fmt.Errorf("attribute values are different, got %s and %s", testStringValue(i), testStringValue(j))
+		}
+
+		return nil
+	}
+}
+
+// testStringValue returns string values from string pointers, handling nil pointers.
+func testStringValue(sPtr *string) string {
+	if sPtr == nil {
+		return ""
+	}
+
+	return *sPtr
+}
+
+// protoV5ProviderFactories returns a muxed ProviderServer that uses the provider code from this repo (SDK and plugin-framework).
+// Used to set ProtoV5ProviderFactories in a resource.TestStep within an acceptance test.
+func protoV5ProviderFactories(t *testing.T) map[string]func() (tfprotov5.ProviderServer, error) {
+	return map[string]func() (tfprotov5.ProviderServer, error){
+		"google": func() (tfprotov5.ProviderServer, error) {
+			provider, err := MuxedProviders(t.Name())
+			return provider(), err
+		},
+	}
+}
+
+// providerVersion450 returns information allowing tests to download TPG v4.50.0 from the Registry during `init`
+// Used to set ExternalProviders in a resource.TestStep within an acceptance test.
+func providerVersion450() map[string]resource.ExternalProvider {
+	return map[string]resource.ExternalProvider{
+		"random": {
```

Comment:
Should this be "google"?

Reply:
Ah copy-paste issue between TPG and MM during development - thanks